### PR TITLE
Refactor map

### PR DESCRIPTION
**Pull request by duncanspumpkin**

### CODE REVIEW

File: src/openrct2/actions/MazeSetTrackAction.hpp [hunk 1]
```diff
@@ -94,7 +94,7 @@ DEFINE_GAME_ACTION(MazeSetTrackAction, GAME_COMMAND_SET_MAZE_TRACK, GameActionRe
             return res;
         }
 
-        if (!map_is_location_owned(_loc) && !gCheatsSandboxMode)
+        if (!map_is_location_owned({ _loc.x, _loc.y, _loc.z }) && !gCheatsSandboxMode)
```
Comment (Gymnasiast) [inline src/openrct2/actions/MazeSetTrackAction.hpp]: Wasn't the plan to allow decaying CoordsXYZD into CoordsXYZ and likewise?

Comment (duncanspumpkin) [reply src/openrct2/actions/MazeSetTrackAction.hpp]: Yes. I haven't got around to doing that though

Comment (Gymnasiast) [reply src/openrct2/actions/MazeSetTrackAction.hpp]: Do you plan to tackle that in this PR or in another?

Comment (duncanspumpkin) [reply src/openrct2/actions/MazeSetTrackAction.hpp]: Can't decide. I'll run a few tests this weekend

File: src/openrct2/actions/BannerPlaceAction.hpp [hunk 2]
```diff
@@ -72,7 +72,7 @@ DEFINE_GAME_ACTION(BannerPlaceAction, GAME_COMMAND_PLACE_BANNER, GameActionResul
             return MakeResult(GA_ERROR::INVALID_PARAMETERS, STR_CANT_POSITION_THIS_HERE, STR_CAN_ONLY_BE_BUILT_ACROSS_PATHS);
         }
 
-        if (!map_can_build_at(_loc.x, _loc.y, _loc.z))
+        if (!map_can_build_at({ _loc.x, _loc.y, _loc.z }))
```
Comment (duncanspumpkin) [inline src/openrct2/actions/BannerPlaceAction.hpp]: no need for this can pass loc

File: src/openrct2/actions/BannerSetStyleAction.hpp [hunk 3]
```diff
@@ -68,7 +68,7 @@ DEFINE_GAME_ACTION(BannerSetStyleAction, GAME_COMMAND_SET_BANNER_STYLE, GameActi
         res->ExpenditureType = RCT_EXPENDITURE_TYPE_LANDSCAPING;
         res->Position.x = banner->position.x * 32 + 16;
         res->Position.y = banner->position.y * 32 + 16;
-        res->Position.z = tile_element_height(banner->position.x, banner->position.y);
+        res->Position.z = tile_element_height({ res->Position.x, res->Position.y });
```
Comment (duncanspumpkin) [inline src/openrct2/actions/BannerSetStyleAction.hpp]: can pass it position

File: src/openrct2/actions/BannerSetStyleAction.hpp [hunk 4]
```diff
@@ -118,7 +118,7 @@ DEFINE_GAME_ACTION(BannerSetStyleAction, GAME_COMMAND_SET_BANNER_STYLE, GameActi
...
```
Comment (duncanspumpkin) [inline src/openrct2/actions/BannerSetStyleAction.hpp]: as above

File: src/openrct2/actions/FootpathPlaceAction.hpp [hunk 5]
```diff
@@ -252,12 +251,11 @@ DEFINE_GAME_ACTION(FootpathPlaceAction, GAME_COMMAND_PLACE_PATH, GameActionResul
             return MakeResult(GA_ERROR::DISALLOWED, STR_CANT_BUILD_FOOTPATH_HERE, STR_CANT_BUILD_THIS_UNDERWATER);
         }
 
-        auto tileElement = map_get_surface_element_at({ _loc.x, _loc.y });
-        if (tileElement == nullptr)
+        auto surfaceElement = map_get_surface_element_at({ _loc.x, _loc.y });
```
Comment (duncanspumpkin) [inline src/openrct2/actions/FootpathPlaceAction.hpp]: pass _loc

File: src/openrct2/actions/FootpathPlaceAction.hpp [hunk 6]
```diff
@@ -314,12 +312,11 @@ DEFINE_GAME_ACTION(FootpathPlaceAction, GAME_COMMAND_PLACE_PATH, GameActionResul
 
         gFootpathGroundFlags = gMapGroundFlags;
 
-        auto tileElement = map_get_surface_element_at({ _loc.x, _loc.y });
-        if (tileElement == nullptr)
+        auto surfaceElement = map_get_surface_element_at({ _loc.x, _loc.y });
```
Comment (duncanspumpkin) [inline src/openrct2/actions/FootpathPlaceAction.hpp]: again

File: src/openrct2/actions/FootpathSceneryPlaceAction.hpp [hunk 7]
```diff
@@ -53,13 +53,12 @@ DEFINE_GAME_ACTION(FootpathSceneryPlaceAction, GAME_COMMAND_PLACE_FOOTPATH_SCENE
...
         res->ExpenditureType = RCT_EXPENDITURE_TYPE_LANDSCAPING;
         res->Position = _loc;
-        if (!map_is_location_valid({ _loc.x, _loc.y }))
+        if (!map_is_location_valid(_loc))
         {
             return MakeResult(GA_ERROR::INVALID_PARAMETERS, STR_CANT_POSITION_THIS_HERE, STR_OFF_EDGE_OF_MAP);
         }
 
-        if (!((gScreenFlags & SCREEN_FLAGS_SCENARIO_EDITOR) || gCheatsSandboxMode)
-            && !map_is_location_owned(_loc.x, _loc.y, _loc.z / 8))
```
Comment (Gymnasiast) [inline src/openrct2/actions/FootpathSceneryPlaceAction.hpp]: You discarded the `/ 8` here.  In essence, this is correct (a function should either take big or small coordinates), but the calls to `map_is_location_owned()` will need to be carefully checked  to see if all of them now pass the `z` coordinate correctly.

File: src/openrct2/ride/Ride.cpp [hunk 8]
```diff
@@ -3239,28 +3239,17 @@ void ride_check_all_reachable()
...
     if (coordinates.isNull())
-        return 1;
+        return true;
 
-    x = coordinates.x;
-    y = coordinates.y;
-    z = coordinates.z;
-    uint8_t face_direction = coordinates.direction;
+    TileCoordsXYZ loc{ coordinates.x, coordinates.y, coordinates.z };
+    loc -= TileDirectionDelta[coordinates.z];
```
Comment (Gymnasiast) [inline src/openrct2/ride/Ride.cpp]: Shouldn't this use `coordinates.direction` instead of `coordinates.z`?

Comment (duncanspumpkin) [reply src/openrct2/ride/Ride.cpp]: Doh

File: src/openrct2/actions/FootpathSceneryPlaceAction.hpp (hunk 7 above)
Comment (duncanspumpkin) [reply src/openrct2/actions/FootpathSceneryPlaceAction.hpp]: Yes I'm pretty confident I have all of these correct.

**Pull request merged**